### PR TITLE
Avoid scatter for single-device case in DDP

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46304 Avoid scatter for single-device case in DDP**

In the case that a single process operates only on one GPU, we can
avoid this scatter and instead replace it with a recursive version of `to`
which transfers the input tensors to the correct device.

The implementation of `_recursive_to` is modeled after `scatter` in https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/scatter_gather.py, in order to keep parity with the previous conventions (i.e. custom types not having their tensors moved).

Differential Revision: [D24296377](https://our.internmc.facebook.com/intern/diff/D24296377/)

### CODE REVIEW

File: torch/testing/_internal/distributed/distributed_test.py [hunk 1]
```diff
@@ -3781,6 +3781,56 @@ def forward(self, x):
...
+                    # Similar to scatter, the recursive to in the single-device
+                    # case does not move tensors if they are in a custom type.
+                    self.assertTrue(isinstance(x, expected_type))
+                    if expected_type == TensorWrapper:
+                        self.assertEqual(str(x.t.device), "cpu")
+                        x.t = x.t.to(self.rank)
+                        return _self.lin(x.t)
+                    else:
+                        self.assertTrue(len(x), expected_len)
+                        self.assertTrue(x[0].device == x[1].device)
```
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/distributed_test.py:3813]: Shouldn't we pass in the expected device to the constructor of ToyModel and validate it is correct here?

File: torch/testing/_internal/distributed/distributed_test.py [hunk 2]
```diff
@@ -3781,6 +3781,56 @@ def forward(self, x):
...
+                ToyModel().to(self.rank), device_ids=[self.rank]
+            )
+            # CPU input, should be moved to the proper device before call to
+            # forward.
+            inp = tuple(torch.randn(10, 10) for _ in range(expected_len))
+            model(inp, tuple)
+            # List CPU input, should be moved to proper device before call to
+            # forward.
+            inp = [torch.randn(10, 10) for _ in range(expected_len)]
+            model(inp, list)
```
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/distributed_test.py:3828]: Can we add tests for dict and namedtuple as well?

File: torch/testing/_internal/distributed/distributed_test.py (hunk 1 above)
Comment (rohan-varma) [reply torch/testing/_internal/distributed/distributed_test.py:3813]: We could do this, but the test is basically expected to validate that the device is the current rank of the process, as we are testing single GPU per process. The next line asserts that the input is on the expected device.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 2 above)
Comment (rohan-varma) [reply torch/testing/_internal/distributed/distributed_test.py:3855]: Added these in the latest version of the diff.

**Review by pritamdamania87 — APPROVED** [review]

File: torch/nn/parallel/distributed.py [hunk 3]
```diff
@@ -660,10 +660,11 @@ def forward(self, *inputs, **kwargs):
             self._check_global_requires_backward_grad_sync(is_joined_rank=False)
 
         if self.device_ids:
-            inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
             if len(self.device_ids) == 1:
+                inputs, kwargs = self.to_kwargs(inputs, kwargs, self.device_ids[0])
                 output = self.module(*inputs[0], **kwargs[0])
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:665]: Curious, any reason we still need to return `inputs` and `kwargs` as a list?

**Review by mrshenli — APPROVED** [review]

File: torch/nn/parallel/distributed.py (hunk 3 above)
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:671]: I was mostly doing that to keep parity with the current version, but we can probably remove it in this code path.

**Comment by ngimel** [comment]
@rohan-varma what's the motivation behind this PR? Does scatter for a single device incur some performance penalty? As #49819 says, previously it was possible to overlap h2d transfers with computation, now since the transfers happen on the default stream (`.to()` uses the default stream, as opposed to side stream in Scatter) this overlap is not possible.
Of course, current implementation can be fixed to use side stream, but I'm wondering if it's worth code complexity.

**Pull request closed without merging**